crates/bevy_render/src/mesh/allocator.rs

```diff
@@ -0,0 +1,1025 @@
+    pending_allocations: HashMap<AssetId<Mesh>, SlabAllocation>,
+
+    /// The layout of a single element (vertex or index).
+    element_layout: ElementLayout,
+
+    /// The size of this slab in slots.
+    slot_capacity: u32,
+}
+
+/// A slab that contains a single object.
+///
+/// Typically, this is for objects that exceed the
+/// [`MeshAllocatorSettings::large_threshold`]. This is also for objects that
+/// would ordinarily receive their own slab but can't because of platform
+/// limitations, most notably vertex arrays on WebGL 2.
+struct LargeObjectSlab {
+    /// The GPU buffer that backs this slab.
+    ///
+    /// This may be `None` if the buffer hasn't been created yet.
+    buffer: Option<Buffer>,
+
+    /// The layout of a single element (vertex or index).
+    element_layout: ElementLayout,
+}
+
+/// The type of element that a slab can store.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+enum ElementClass {
+    /// Data for a vertex.
+    Vertex,
+    /// A vertex index.
+    Index,
+}
+
+/// Information about the size of individual elements (vertices or indices)
+/// within a slab.
+///
+/// Slab objects are allocated in units of *slots*. Usually, each element takes
+/// up one slot, and so elements and slots are equivalent. Occasionally,
+/// however, a slot may consist of 2 or even 4 elements. This occurs when the
+/// size of an element isn't divisible by [`COPY_BUFFER_ALIGNMENT`] (4). When we
+/// resize buffers, we perform GPU-to-GPU copies to shuffle the existing
+/// elements into their new positions, and such copies must be on
+/// [`COPY_BUFFER_ALIGNMENT`] boundaries. Slots solve this problem by
+/// guaranteeing that the size of an allocation quantum is divisible by both the
+/// size of an element and [`COPY_BUFFER_ALIGNMENT`], so we can relocate it
+/// freely.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+struct ElementLayout {
+    /// Either a vertex or an index.
+    class: ElementClass,
+
+    /// The size in bytes of a single element (vertex or index).
+    size: u64,
+
+    /// The number of elements that make up a single slot.
+    ///
+    /// Usually, this is 1, but it can be different if [`ElementLayout::size`]
+    /// isn't divisible by 4. See the comment in [`ElementLayout`] for more
+    /// details.
+    elements_per_slot: u32,
+}
+
+/// The location of an allocation and the slab it's contained in.
+struct MeshAllocation {
+    /// The ID of the slab.
+    slab_id: SlabId,
+    /// Holds the actual allocation.
+    slab_allocation: SlabAllocation,
+}
+
+/// An allocation within a slab.
+#[derive(Clone)]
+struct SlabAllocation {
+    /// The actual [`Allocator`] handle, needed to free the allocation.
+    allocation: Allocation,
+    /// The number of slots that this allocation takes up.
+    slot_count: u32,
+}
+
+/// Holds information about all slabs scheduled to be allocated or reallocated.
+#[derive(Default, Deref, DerefMut)]
+struct SlabsToReallocate(HashMap<SlabId, SlabToReallocate>);
```

Comment:
HashMap seems excessive here and in SlabToReallocate. Do we have some kind of SmallVec-like map type?

Reply:
No, we don't, and I don't think it's worth pulling a dependency and bloating Bevy's code size to optimize this. Reallocation is super slow no matter what. The overhead of a couple of CPU side allocations are a drop in the bucket next to allocating the buffer on GPU and copying over.